implementations/rust/ockam/ockam_entity/src/profile_set.rs

```diff
@@ -11,62 +11,125 @@ pub use identifiers::*;
-    profiles: Vec<P>,
+    profiles: HashMap<ProfileIdentifier, P>,
+}
+
+pub trait ProfileRetrieve<P: ProfileTrait> {
+    fn profile(&self, profile_identifier: &ProfileIdentifier) -> Option<&P>;
+}
+
+pub trait ProfileAdd<P: ProfileTrait> {
+    fn add_profile(&mut self, profile: P) -> Result<()>;
+}
+
+pub trait ProfileUpdate<P: ProfileTrait> {
+    fn update_profile(&mut self, old_profile_id: &ProfileIdentifier, profile: P) -> Result<()>;
 }
 
-impl<P: ProfileTrait> Entity<P> {
+pub trait ProfileRemove {
+    fn remove_profile(&mut self, profile_id: &ProfileIdentifier) -> Result<()>;
+}
+
+pub trait ProfileManagement<P: ProfileTrait>:
+    ProfileRetrieve<P> + ProfileAdd<P> + ProfileUpdate<P> + ProfileRemove
+{
+}
+
+impl<P: ProfileTrait, U> ProfileManagement<P> for U where
+    U: ProfileRetrieve<P> + ProfileAdd<P> + ProfileUpdate<P> + ProfileRemove
+{
+}
+
+impl<P: ProfileTrait> ProfileSet<P> {
     /// Create a new Entity with the given default profile.
     pub fn new(default_profile: P) -> Self {
         let idref = default_profile.identifier().unwrap();
         let default_profile_identifier = ProfileIdentifier::from_key_id(idref.key_id().clone());
-        let profiles = vec![default_profile];
-        Entity {
+
+        let mut profiles = HashMap::new();
+        profiles.insert(default_profile_identifier.clone(), default_profile);
+
+        ProfileSet {
             default_profile_identifier,
             profiles,
         }
     }
 
     fn default_profile(&self) -> Option<&P> {
-        self.profiles
-            .iter()
-            .find(|profile| self.default_profile_identifier == profile.identifier().unwrap())
+        self.profile(&self.default_profile_identifier)
     }
+}
+
+impl<P: ProfileTrait> ProfileAdd<P> for ProfileSet<P> {
+    fn add_profile(&mut self, profile: P) -> Result<()> {
+        if let Ok(id) = profile.identifier() {
+            if let Some(_) = self.profiles.insert(id, profile) {
+                return Ok(());
+            }
+        }
+        Err(InvalidInternalState.into())
+    }
+}
+
+impl<P: ProfileTrait> ProfileRetrieve<P> for ProfileSet<P> {
+    fn profile(&self, profile_identifier: &ProfileIdentifier) -> Option<&P> {
+        self.profiles.get(profile_identifier)
+    }
+}
 
-    fn _add_profile(&mut self, profile: P) {
-        self.profiles.push(profile);
+impl<P: ProfileTrait> ProfileRemove for ProfileSet<P> {
+    fn remove_profile(&mut self, profile_id: &ProfileIdentifier) -> Result<()> {
+        if &self.default_profile_identifier == profile_id {
+            return Err(InvalidParameter.into());
+        }
+        if let Some(_) = self.profiles.remove(&profile_id) {
+            Ok(())
+        } else {
+            Err(InvalidInternalState.into())
+        }
+    }
+}
+
+impl<P: ProfileTrait> ProfileUpdate<P> for ProfileSet<P> {
+    fn update_profile(&mut self, old_profile_id: &ProfileIdentifier, profile: P) -> Result<()> {
+        self.remove_profile(&old_profile_id)?;
```

Comment:
That should be definitely improved in the future by applying new changes to existing profile. But I think it's good for now, but let's have an issue or at least comment in the code

Reply:
Ehh, given your other comments regarding change events, I think this function is premature. Going to remove.